Code/Framework/AzCore/AzCore/IO/Streamer/RecentlyUsedIndex.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/std/functional.h>
+#include <AzCore/std/numeric.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <AzCore/std/typetraits/is_integral.h>
+
+namespace AZ::IO
+{
+    //! Uses an optimized double linked list to keep track of the most and least recently used index.
+    template<typename T>
+    class RecentlyUsedIndex
```

Comment:
Curious what the reasoning was to make this a template instead of a concrete class using an underlying u32 type?  It looks like the only place it is not used as a u32 type is in the tests and having this type in IO/Streamer makes it seem specific to Streamer and unlikely to be used elsewhere.  Main concern is build time and complexity of using templating.

Reply:
Cache efficiency, if a 16-bit value is used twice as much fits in a single cache line. The default would probably best be a `size_t` but using a 32-bit value give better results. Probably could go 16-bit in some places, so this allows for experimenting with that more easily.